File: spark-jobs/src/main/scala/com/griddynamics/training/spark/jobs/IpSubnetMask.scala

package com.griddynamics.training.spark.jobs

import java.net.{InetAddress, UnknownHostException}

@SerialVersionUID(200L)
class IpSubnetMask(private val address: Int, private val mask: Int, val getGeoNameId: Int) extends Serializable {

  def matches(addressToCheck: Int): Boolean = (address & mask) == (addressToCheck & mask)

  override def toString = s"IpSubnetMask($address, $mask, $getGeoNameId)"
}

@SerialVersionUID(201L)
object IpSubnetMask extends Serializable {
  def apply(subnet: String, geoNameId: Int): IpSubnetMask = {
    val parts: Array[String] = subnet.split("/")
    val ip = parts(0)
    val prefix = if (parts.length < 2) 0 else parts(1).toInt

    new IpSubnetMask(ipAddressToInt(ip), -1 << (32 - prefix), geoNameId)
  }

  def ipAddressToInt(ipAddress: String): Int = {
    var a1: Option[InetAddress] = Option.empty
    try {
      a1 = Option(InetAddress.getByName(ipAddress))
    } catch {
      case _: UnknownHostException =>
    }
    a1.map(_.getAddress)
      .map(b1 => ((b1(0) & 0xFF) << 24) | ((b1(1) & 0xFF) << 16) | ((b1(2) & 0xFF) << 8) | ((b1(3) & 0xFF) << 0))
      .getOrElse(0)
  }
}

File: custom.hive.udf/src/main/java/com/griddynamics/training/vk/hive/udf/IpSubnetMask.java
package com.griddynamics.training.vk.hive.udf;

import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.Optional;

public class IpSubnetMask {

    private int address;
    private int mask;
    private int geoNameId;

    public IpSubnetMask(int[] address, int prefix, int geoNameId) {
        this.address = ipAddressToInt(address);
        this.mask = -(1 << (32 - prefix));
        this.geoNameId = geoNameId;
    }

    public static Optional<Integer> ipAddressToInt(String ipAddress) {
        return getIpAddressSegments(ipAddress).map(IpSubnetMask::ipAddressToInt);
    }

    public static int ipAddressToInt(int[] address) {
        return address[0] << 24 | address[1] << 16 | address[2] << 8 | address[3];
    }

    public static Optional<int[]> getIpAddressSegments(String ipAddress) {
        try {
            int[] ipSegments = new int[4];
            int index = 0;
            for (byte b : InetAddress.getByName(ipAddress).getAddress()) {
                ipSegments[index++] = b & 0xFF;
            }
            return Optional.of(ipSegments);
        } catch (UnknownHostException e) {
            return Optional.empty();
        }
    }

    public boolean matches(int addressToCheck) {
        return (address & mask) == (addressToCheck & mask);
    }

    public int getGeoNameId() {
        return geoNameId;
    }

    @Override
    public String toString() {
        return "IpSubnetMask{" + "address=" + address +
                ", mask=" + mask +
                ", geoNameId=" + geoNameId +
                '}';
    }
}
